File: GPT-4/Stack/Hard/Runtime and Memory/736. Parse Lisp Expression/3.cpp
#include <string>
#include <unordered_map>
#include <sstream>

using namespace std;

class Solution {
public:
    int evaluate(string expression) {
        unordered_map<string, int> scope;
        return evaluate(expression, scope);
    }
    
private:
    int evaluate(const string& expression, unordered_map<string, int>& scope) {
        if (isdigit(expression[0]) || expression[0] == '-')
            return stoi(expression);
        
        if (isalpha(expression[0]))
            return scope[expression];
        
        int i = 1, j;
        string cmd = parse(expression, i);
        if (cmd == "add") {
            int left = evaluate(parse(expression, i), scope);
            int right = evaluate(parse(expression, i), scope);
            return left + right;
        } else if (cmd == "mult") {
            int left = evaluate(parse(expression, i), scope);
            int right = evaluate(parse(expression, i), scope);
            return left * right;
        } else { // let
            unordered_map<string, int> newScope = scope;
            while (true) {
                string var = parse(expression, i);
                if (i >= expression.size())
                    return evaluate(var, newScope);
                string value = parse(expression, i);
                newScope[var] = evaluate(value, newScope);
            }
        }
    }
    
    string parse(const string& expression, int& i) {
        if (i >= expression.size()) return "";
        
        int j = i;
        while (i < expression.size() && expression[i] != ' ' && expression[i] != ')') i++;
        string res = expression.substr(j, i - j);
        while (i < expression.size() && (expression[i] == ' ' || expression[i] == ')')) i++;
        return res;
    }
};

File: Java/GPT-4/Stack/Hard/Runtime and Memory/736. Parse Lisp Expression/3.java
import java.util.*;

class Solution {
    public int evaluate(String expression) {
        return eval(expression, new ArrayDeque<>());
    }

    private int eval(String exp, Deque<Map<String, Integer>> scope) {
        if (Character.isDigit(exp.charAt(0)) || exp.charAt(0) == '-')
            return Integer.parseInt(exp);
        else if (!exp.startsWith("(")) {
            Map<String, Integer> currentScope = scope.peek();
            if (currentScope.containsKey(exp)) {
                return currentScope.get(exp);
            } else {
                return 0;  // Return a default value for undefined variables
            }
        }

        // Create a new scope
        Map<String, Integer> map = new HashMap<>();
        scope.push(map);
        
        // To handle variable and function expressions separately
        String[] tokens;
        if (exp.charAt(1) == 'l') {  // let expression
            tokens = exp.substring(5, exp.length() - 1).split(" ");
        } else {  // add or mult expression
            tokens = exp.substring(6, exp.length() - 1).split(" ");
        }

        int res = 0;
        if (exp.charAt(1) == 'a')  // add expression
            res = eval(tokens[0], scope) + eval(tokens[1], scope);
        else if (exp.charAt(1) == 'm')  // mult expression
            res = eval(tokens[0], scope) * eval(tokens[1], scope);
        else {  // let expression
            for (int i = 0; i < tokens.length - 2; i += 2)
                map.put(tokens[i], eval(tokens[i + 1], scope));
            res = eval(tokens[tokens.length - 1], scope);
        }

        // Remove the scope
        scope.pop();
        return res;
    }
}

